### ISSUE TITLE
Adding/Deleting resource scopes and application grants at the same time creates a race condition

### ISSUE THREAD

**Issue by dbryar** [pingidentity/terraform-provider-pingone#774]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

Thank you for opening an issue. Please note that we try to keep the Terraform issue tracker reserved for bug reports and feature requests. For general usage questions, please see: https://www.terraform.io/community.html.

### PingOne Terraform provider Version

0.26

### Terraform Version

1.4.2

### Affected Resource(s)

- pingone_application_resource_grant
- pingone_resource_scope

Changing resource scopes and grants in the same MR creates a race condition;

### Expected Behaviour

I was expected that the provider would enforce an order of events in the changes to resources, scopes and grants given they are specifically required to be respected in order to create, update and delete.

- When creating scopes should be asynchronously added as a resource, then applied as grants to applications
- When removing grants _and_ deleting scopes the grants should be asynchronously removed from the application and then the scopes (and resources) modified

Terraform should observe dependencies
- It should not be able to delete scopes from resources that are in use by any application
- It should not be able to modify grants to applications where the scope is in a state of change

### Actual Behaviour

Scopes were partially removed from the resource
Grants were partially modified in the application
Application grants were no longer accessible via Ping UI (console)
![Screenshot 2024-04-04 at 2 43 01 pm](https://github.com/pingidentity/terraform-provider-pingone/assets/20749904/642cfe06-89c9-4576-8d24-c1f9f5fb9781)

### Steps to Reproduce

1. create an application with no specific scopes
2. apply changes
3. merge an MR that creates a resource and add a scope
4. apply changes
5. merge an MR that adds the scope grant to an application
6. apply changes
7. revert code to match step 1
8. apply changes

**Comment by dbryar** [comment]
Using the Ping API it was possible to read all grants from the application showing a list of grants, two of which had IDs that no longer existed

Deleting the grant via Ping API was the only way to fix the application

**Comment by patrickcping** [comment]
Hi @dbryar thanks for raising!  We'll need a little time to work through this one.  It looks like you've been able to work around the issue for now?

**Comment by dbryar** [comment]
Just ensuring dependency order is respected in MRs for now

**Comment by patrickcping** [comment]
hey @dbryar , I can't replicate the issue from the steps provided.  In my tests, the dependency order in returning to step 1 seems to be respected and there are no console UI issues.

In my tests, the dependency order may not be respected with the following syntax (notice that the `pingone_application_resource_grant` uses hardcoded text values, or values that could come from a variable), but the API responds correctly by revoking the resource scope from the application automatically.

```hcl

// 1 and 2
resource "pingone_application" "my_awesome_web_app" {
  environment_id = pingone_environment.my_environment.id
  name           = "My Awesome Web App"
  enabled        = true

  oidc_options {
    type                        = "WEB_APP"
    grant_types                 = ["AUTHORIZATION_CODE", "REFRESH_TOKEN"]
    response_types              = ["CODE"]
    token_endpoint_authn_method = "CLIENT_SECRET_BASIC"
    redirect_uris               = ["https://my-website.com"]
  }
}

// 3 and 4
resource "pingone_resource" "my_resource" {
  environment_id = pingone_environment.my_environment.id

  name = "My resource"
}

resource "pingone_resource_scope" "my_resource_scope" {
  environment_id = pingone_environment.my_environment.id
  resource_id    = pingone_resource.my_resource.id

  name = "example_scope"
}

// 5 and 6
resource "pingone_application_resource_grant" "my_awesome_spa_custom_resource_grants" {
  environment_id = pingone_environment.my_environment.id
  application_id = pingone_application.my_awesome_web_app.id

  resource_name = "My resource" // for correct ordering, this should instead be pingone_resource.my_resource.name

  scope_names = [
    "example_scope", // for correct ordering, this should instead be pingone_resource_scope.my_resource_scope.name
  ]
}
```

Are you able to please verify your `pingone_application_resource_grant` isn't using hardcoded values for the resource name and scopes, and share minimal HCL for steps 1, 3, 5 and 7 that I can verify?

**Comment by dbryar** [comment]
It hasn't come up again since so I can't absolutely confirm the order of events, but it is probably the way our code is generating the HCL as its now a JSON from CDK-TF

All the resource grants are bundled in to the one block, for the entire deploy, e.g.
```json
{
  "pingone_application_resource_grant": {
    "acceptance_e2e_tests_acceptance": {
      "application_id": "${pingone_application.acceptance_e2e_tests.id}",
      "environment_id": "${pingone_environment.dev.id}",
      "resource_name": "${pingone_resource.acceptance.name}",
      "scope_names": ["consent", "admin.consent.read", "admin.consent.write", "admin.agreement.read", "admin.agreement.write"]
    },
    "acceptance_e2e_tests_openid": {
      "application_id": "${pingone_application.acceptance_e2e_tests.id}",
      "environment_id": "${pingone_environment.dev.id}",
      "resource_name": "${data.pingone_resource.openid.name}",
      "scope_names": ["profile", "email", "phone"]
    }
  }
}
```

**Comment by dbryar** [comment]
It has happened again simply changing the scope string name.

Applying the change worked in all lower envs and then broke in prod deploy.

The fix is to DELETE the grant for the application and reapply Terraform

It is believed the issue is when the scope text is used as the terraform ID so when it changes, terraform deletes the "old" scope and then creates a "new" scope. If the ID was consistent this would not occur as the link is done via object ID, and the name change should return the same ID for both.

**Comment by patrickcping** [comment]
Took me a while, but got the problem reliably recreated by renaming the resource that the scopes belong to (triggering a mass simultaneous replacement of all the resource scopes).  The issue only occurs when the resource scopes are granted to an application, the blocking errors are from the platform's own referential integrity as it updates the application grants in the background.

I'll raise with the platform team to see how they want to handle it in the API, but there are things we can do in the provider code to remove scopes sequentially rather than all at once.

The following HCL reproduces the issue by first applying, switching `flip` to `false` and applying again

```terraform
locals {
  flip = true

  resources = [
    local.flip ? "My resource" : "My resource2",
    local.flip ? "My resource2" : "My resource"
  ]

  scopes = [
    "example_scope1",
    "example_scope2",
    "example_scope3",
    "example_scope4",
    "example_scope5"
  ]
}

resource "pingone_resource" "my_resource" {
  count = length(local.resources)

  environment_id = pingone_environment.my_environment.id
  name           = local.resources[count.index]
}

resource "pingone_resource_scope" "my_resource_scope" {
  for_each = toset(local.scopes)

  environment_id = pingone_environment.my_environment.id
  resource_id    = pingone_resource.my_resource[local.flip ? 0 : 1].id
  name           = each.value
}

resource "pingone_application_resource_grant" "my_awesome_spa_custom_resource_grants" {
  environment_id = pingone_environment.my_environment.id
  application_id = pingone_application.my_awesome_web_app.id

  resource_name = pingone_resource.my_resource[local.flip ? 0 : 1].name
  scope_names   = local.scopes

  depends_on = [pingone_resource_scope.my_resource_scope]
}
```

### Error

```
│ Error: Error when calling `DeleteResourceScope`: The request could not be completed. One or more validation errors were in the request.
│ 
│ PingOne Error Details:
│ ID: cd1b****-****-****-****-********389b
│ Code: INVALID_DATA
│ Message: The request could not be completed. One or more validation errors were in the request.
│ Details object: [{"code":"INVALID_VALUE","message":"Unable to find scopes with the following IDs: 92a0****-****-****-****-********7ea8,
│ f24f****-****-****-****-********44ca","target":"scopes"}]
```

**Comment by patrickcping** [comment]
The fix for this issue was released today in version `v0.29.1`.

I'll shortly be releasing `v1.0.0-rc1`, a pre-release candidate build which is not yet ready for production use.  There is no requirement to upgrade from `v0.*` to `v1.*` to get this fix, as customers can stay on `v0.29.1` until ready to upgrade after the full `v1.0.0` release is available.
